### DISCUSSION TITLE
What could be fixed in Security in Moonwalk?

### DISCUSSION THREAD

**Discussion by darrelmiller** [OAI/sig-moonwalk#46] [General]
- Workflows is trying to describe token acquisition sequences.  What can we do to help?
- Attach scopes to operation + subsets of a model
- Align with the security SIG
-

**Comment by darrelmiller** [comment]
@whitlockjc Add your thoughts here please!

**Comment by whitlockjc** [comment]
[Security Schemes](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#security-scheme-object) _(formerly known as "Security Definitions")_ provide a way to describe the security scheme(s) used by your API. Unfortunately, the _types_ supported by OpenAPI are not only limited but they often tie the credential location to the Security Scheme type.  This proposal will add a generic way to define where a credential is and its format so that the consumer can have complete control over where credentials are used in their API.  Not only would this allow consumers to specify their own Security Schemes without OpenAPI having to be the gatekeeper, but it also allows a level of flexibility not available now.  There have also been extra Security Scheme properties that have been _tacked on_ to support the growing needs of consumers, which would be alleviated _(or moved in this case)_.

**Note:** Sourced from the _Motivation_ section of the [original proposal](https://github.com/OAI/OpenAPI-Specification/pull/2582 ).

**Comment by darrelmiller** [comment]
Stu:  How do influence the community to create more consistent implementations?  Lots of variations of OAuth2.  
Jeremy: Discussed the proposed solution.  Cleaned up structure with a config and credentials object. Also made `type` extensible.
Dan: We need a better model for Authorization.
Marsh: Does this need to be part of OpenAPI? Could this be used for other scenarios? What are the scenarios it is used for?
Darrel: Discovery of how to provision an api key, or an OAuth2 client application?
Marsh: Other metadata, how long do tokens last? How often do I need to rotate secrets?
Dan: Supportive of considering externalizing schema.  Also, could security events be described in a different document.
MikeR: AsyncAPI has the notion of bindings for protocol specific information that defer to an external document.
Dan: Multiple documents forces the separation of concerns. Would we need to support embedding to help customers who really want everything in one file.
Ralf: likes the idea of separating AuthN and API Design
MikeR: + 1
Darrel:  Is AuthZ part of the API design or Security concern?
Dan says yes, Ralf says no.
Ben: What are the concerns of an OAS document?   
MikeR: Pulling security (and potentially other areas) out of "core" might enable us to iterate on moonwalk faster

**Comment by mcrobbj** [comment]
The security section assume that there is just a single Idp across all the environments. You can only define a single authorisation and token endpoint. Could it be like the servers section for the API so that you can have environment specific idp endpoints.

Also there is mention of mtls being able to be defined but I have never managed to find an example. of how to use it in the operations level. So I only have the following for the security schema:
     mutualTLS:
      description: Mutual TLS
      type: mutualTLS
      scheme: mutual

**Comment by rafalkrupinski** [comment]
> subsets of a model

Is that something like readOnly/writeOnly from OAI 3.0?

**Reply by orubel** [reply]
No. ROLES (as I repeatedly talked about since 2015) are supported as a standard (see RBAC/ABAC). As a standard, the backend supplies a DEFAULT of 'ROLE_ANONYMOUS' when a user has no role found or no associated JWT. Aside from that, ROLES are aside at the PARENT level to endpoints:

https://gist.github.com/orubel/6d529d39133e39b3f28d7cf1b9240236

Hopefully, this time around, we can get a better response than this : https://flic.kr/p/2keNR8v

Additional, this is NOT attached to 'model'... this is an association with 'endpoint' ("Attach scopes to operation + subsets of a model"). The endpoint is attache to business logic which is a RequestHandler (see AWS, SpringBoot, Flask, etc). This is what a controller extends and hence where you endpoint resides (regardless of REACTOR pattern or FRONT CONTROLLER pattern)

You can see ROLES attached via FLASK/SPRINGBOOT annotations and even in Django at the controllers (Django calls CONTROLLERS 'views')

**Comment by handrews** [comment]
3.x issue move to Moonwalk:

* https://github.com/OAI/OpenAPI-Specification/issues/2628

**Comment by mikekistler** [comment]
Hi folks, I want to chime in here with another perspective. I think it would be really nice if OpenAPI security schemes could describe "real world" authn/authz flows like cookie authentication. I chatted about this recently with @darrelmiller and we concluded that the only way to describe this in OpenAPI 3.x is with:
```yaml
components:
  securitySchemes:
    cookieAuth:
      type: apiKey
      in: cookie
      name: token
```

which to me seems totally inadequate. 

For example, it does not indicate what operation is used to obtain the cookie -- typically this is something like "/login". 

Also, "apiKey" seems wrong. I think of getting an API key from the service through some out-of-band channel.
Cookies are nothing like that.

Does anyone else think that this is a problem that needs fixing in Moonwalk?

**Reply by orubel** [reply]
well again, token payload stores ROLES for RBAC:
```
{
  "id": "some id",
  "exp": 1602522810,
  "app_metadata": {
    "authorization": {
      "roles": ["admin","editor"]
    }
  }
  "iat": 1607022935
}
```

Those need to be directly associated with endpoints for RBAC/ABAC for all request/response call flow especially in the case of internal/external; redirection (see Arazzo convo - https://github.com/OAI/Arazzo-Specification/discussions/23)

MANY people forget that HTTP standard uses an internal redirect as well an external one; they thing just because REACTOR abstracts front controller, all they need to worry about is external. Highly scalable frameworks like Spring/springboot and other multithreaded use PROACTOR and are fully HTTP compliant and thus can use an internal loopback (see forward vs redirect)

Plus AWS has implemented STEP FUNCTIONS as a bandaid fix to be HTTP COMPLIANT

So you still need to implement ROLES internally for internal redirects

**Reply by SensibleWood** [reply]
@mikekistler I agree that Moonwalk could facilitate doing a more comprehensive "job" in this area. The Security Schemes range from very detailed (OAuth Flows), providing useful pointers (OpenID Discovery), to basic signposting (HTTP).

It would great to get some consistency across the level of detail provided by each Security Scheme. In case the Security Scheme Object should do enough to point the API consumer to what they need to "do" to take the relevant steps to invoke an operation with the relevant credentials and entitlements.

To me this would significantly enhance usability, and provide affordances for tooling makers implementing OpenAPI.

**Reply by orubel** [reply]
Yeah I mentioned this to Tony Tam back in 2015 and [I REPEATEDLY mentioned it to Darrel Miller](https://flic.kr/p/2keNR8v)

This is something they have deliberately ignored for almost 10 years now.

**Reply by ralfhandl** [reply]
We could cautiously add to the Security Schemes in 3.2.0 to get more consistency.

@orubel, @SensibleWood  Would you like to open pull requests for 3.2.0?

Roles and RBAC probably are too big a change for a minor version and are better tackled in Moonwalk.

**Reply by orubel** [reply]
@ralfhandl would love to but the issue arises that even if you define in Moonwalk, the openapi definition defines an endpoint free of RBAC/ABAC definitions which then is contradictory. How would that work? How would that remain consistent?

Plus one would have to use the definitions/values defined in OpenApi. Would this require a level of redundancy to remain consisitent?

**Reply by ralfhandl** [reply]
@orubel I don't understand where the contradiction/inconsistency would come from. Moonwalk _is_ OpenAPI 4.0. And if an API description contradicts the API implementation, that can happen independently of the OpenAPI version used to produce the inaccurate description.

**Reply by orubel** [reply]
@ralfhandl My apologies for misunderstanding. I thought moonwalk was a complimentary security schema.

Maybe I can write a push to suggest a change for RBAC/ABAC to be in compliance. I'll take a quick look over the spec and see what I can suggest.

**Reply by ralfhandl** [reply]
@orubel The Moonwalk project hasn't yet gotten to the point of writing a draft spec version yet, we are still in the discussion phase. All we currently have is a [preliminary outline](https://github.com/OAI/sig-moonwalk/blob/main/specification/moonwalk-source.md), and you could open a PR trying to fit your proposal into one or more of the current section, or propose new sections.

Or you could  add to this discussion, or spawn a related discussion on RBAC/ABAC/...

**Reply by SensibleWood** [reply]
> We could cautiously add to the Security Schemes in 3.2.0 to get more consistency.
> 
> @orubel, @SensibleWood Would you like to open pull requests for 3.2.0?
> 
> Roles and RBAC probably are too big a change for a minor version and are better tackled in Moonwalk.

@ralfhandl one area that keeps coming back to haunt me is JSON Web Signatures and JSON Web Encryption. It would be **_so_** great to be able to wrap up the JWT attributes and payload into a concise and easy-to-use definition. The payload bit is fine - that's just a Scheme Object - but having something like a `header` structure to accompany the payload definition. Ideally you'd provide it as an adjunct to the payload:

e.g.

```
{
  "header": { "$ref": "/components/securitySchemes/JweHeader" },
  "payload": { "$ref": "/components/schemas/JwePayload" }
}
```

Getting this nailed would be so great because we basically be able to _accurately_ describe every open finance standard out there and all the supporting security standards. At the moment it's all "implied" i.e. standards authors write the payload structure and then either a load of description or a Spec Extension.

I'll stroll over to the 3.2 repo and see what I can see. @handrews I believe there was an issue out there on JWT support but I can't find it right now...

**Reply by ralfhandl** [reply]
> here was an issue out there on JWT support

Was it
* https://github.com/OAI/OpenAPI-Specification/issues/1971

Or one of these: https://github.com/OAI/OpenAPI-Specification/issues?q=is%3Aissue+is%3Aopen+jwt

**Reply by orubel** [reply]
@SensibleWood Well RBAC/ABAC are not part of security. They are tightly bound to the endpoint definitions in saying what params to expect with the request/response per role (as you see from above). This needs to be part of the main spec to be able to work.

Separating them out makes them means you will then have a **one definition in OpenAPI  and another in secure definitions which makes no sense**; why do you need that separate definition in openapi??? You are either declaring secure endpoints or you are not.

This has to do with how the schema is thought about. Is it a :
- swagger based testing/apidocs document
- api rules document

You can have both but you have to start thinking of the main doc as an api rules document (not just testing/apidocs), this is a place where all rules can be shared and synchronized and even loaded into a common cache on the backend for a services that share the request/response to see the common rules

The shared state of the request/response has to do 'checks' and the RULES for the checks have to be declared someplace. Often they are hardcoded RULES. But in a distributed architecture, we have to share these rules for the call flow. This means we have to declare them in a schema in order to share them so the request/respoinse has something to compare against for these 'checks'.

But if you define one set of endpoints in OpenAPI and then try to set different endpoint RULES, **you will get a conflict**.

This would require two separate servers and resources: a public api server and a private api server. Thus it would also requires two separate openapi schemas for each server (no one in their right mind is going to try to use one schema for separate servers unless they are clones)

**Reply by SensibleWood** [reply]
Ah it was this puppy: https://github.com/OAI/OpenAPI-Specification/issues/1464 👈 @handrews revisited it recently and pointed out it's already possible with the JSON Schema reference included in  3.1. Looks like the only work to do is get every open finance ecosystem to uplift to 3.1.0...

**Reply by orubel** [reply]
@SensibleWood Yeah I actually use a version of Json Schema to meet my security needs. This is a very old version: https://gist.github.com/orubel/159e94db62023c78a07ebe6d86633763

This also enables me to reset all of these values on the fly without restarting my JAVA server.

**Reply by orubel** [reply]
The thing is, I think there is some confusion on having to STORE the JWT info.

You dont have to. You just need to have the 'ROLES' that it will be sending associated with the endpoints to check for access.

I think this is where the disconnect is.

Plus you can always store in a properties file your networkgrps and the associated roles (as shown below):
`networkRoles: {'open':['ROLE_ADMIN','ROLE_ANONYMOUS','ROLE_USER'],'public':['ROLE_ADMIN','ROLE_USER'],'private':['ROLE_ADMIN']}`

**Reply by SensibleWood** [reply]
@orubel I wasn't making an reference to roles at all in what I was proposing. @ralfhandl asked us where we could contribute, I suggested a place I could, nothing to do with roles.

Describing the shape of a JSON Web Token is important for the reasons I mentioned as it helps people implement standards like [PAR](https://datatracker.ietf.org/doc/html/rfc9126), [JAR](https://www.rfc-editor.org/rfc/rfc9101.html), [OIDC Request Objects](https://openid.net/specs/openid-connect-core-1_0.html#RequestObject), [OIDC ID Tokens](https://openid.net/specs/openid-connect-core-1_0.html#IDToken), etc.

At the moment there's a basic disconnect in the API descriptions for these standards sub version 3.1, namely tooling makers can't pick up the description and reflect the design in a fully-formed Scheme Object. It's all implied in the way I described.

Anyway, looked like v3.1 gives open finance ecosystems what they need.

**Reply by orubel** [reply]
> @orubel I wasn't making an reference to roles at all in what I was proposing. 

gotcha. My apologies for misunderstanding

**Comment by orubel** [comment]
@ralfhandl well I just spent a bit of time on a pull request to have it rejected because I am not a collaborator. 

You can see how springboot does it:

```
    @Secured({"ROLE_1", "ROLE_2"})
    @GetMapping(value = "/accounts/{accountNumber}/reports/DAILY")
    public Report getReportDaily(@PathParam("accountNumber") String accountNumber,
                              @RequestParam("date")Date date){
        reportService = new DailyReportServiceImpl();
        return reportService.getReport();
    }
```

But this is not ideal because **these values are hardcoded**; they cannot be shared in a distributed architecture. They also do not have a separate set of param declarations per role; this has to be hand coded in the application. If params were declared, this can be easilt automated via a service ( I know as I have done this and it is very repeatable in ANY LANGUAGE)

Ideally we want a way to **declare these RULES in a sharable document** so that we can load them at runtime, cache them and apply them prior to calling the business logic/endpoint and after getting the resource ([see interceptor pattern](https://en.wikipedia.org/wiki/Interceptor_pattern)). That way all services have the same logic/security rules for endpoints no matter how it is routed.

To do this we need to declare the roles with the request/response as well as params sent/received with each role:

```
	"show": {
		"METHOD": "GET",
		"DESCRIPTION": "Description for show",
                ...
		"REQUEST": {
			"permitAll": [],
			"ROLE_ADMIN": ["id"]
		},
		"RESPONSE": {
			"permitAll": ["id", "version", "username", "email", "enabled", "accountExpired"],
			"ROLE_ADMIN": ["firstName","lastName"]
		}
	},
```

As you see above, we put the data in the 'endpoint' declaration (as thats where the association exists).

We use a 'permitAll' as a 'catchall' for all ROLES that are not specifically declared but that are given permission to these endpoints in the NETWORKGROUP

I personally declare my NETWORKGRP in a separate properties file but I reference the descriptions in the JSON schema. For example:

`networkRoles: {'open':['ROLE_ADMIN','ROLE_ANONYMOUS','ROLE_USER'],'public':['ROLE_ADMIN','ROLE_USER'],'private':['ROLE_ADMIN']}`

and then in the JSON schema I reference it like so:

`"NETWORKGRP": "public",`


Now why do we do this? Because as we stated above, the JWT passes the ROLE information in the payload and we need someting to check it against:
```
{
  "id": "some id",
  "exp": 1602522810,
  "app_metadata": {
    "authorization": {
      "roles": ["admin","editor"]
    }
  }
  "iat": 1607022935
}
```

If we do not have these in a common shareable document (and these rules associated with the endpoint), the we can have services with different data, different rules and depending on how the request/response is routed, we can end up with a security hole/violation.

**Reply by handrews** [reply]
@orubel we might have limited PRs to keep people from opening PRs on the initial proposal thinking it was the actual draft...  in general, you do not need to be an official collaborator to make PRs here.  Most of the work is happening in discussions right now so it's probably better to continue here or start a new discussion.  The only PRs we're doing right now are for ADRs (Architectural Decision Records, of which we have I think only one very minor one so far).

**Reply by orubel** [reply]
@handrews Thats the least of my concerns. The greater concern is getting this fixed once and for all. Have been pushing for this to get fixed for almost 10 years now. If OpenAPI hopes to be a defacto for security, this is a big one that it is going to have to address.
